### ISSUE TITLE
New appliance - Syncthing

### ISSUE THREAD

**Issue by MPTMG** [turnkeylinux/tracker#897]
If there is interest.... I will start work on a syncthing appliance. I should be able to get this done quickly.

https://syncthing.net

It is a useful tool for keeping files in sync and I did not see it in the Appliance Candidates.

**Comment by JedMeister** [comment]
Hey @MPTMG - that looks like an awesome appliance candidate and we'd love to add it to the library. However, if I were you, I'd be inclined to hold off for a week or 2. 

The rationale behind that is that we're currently pushing out the last of the v14.2 release and I'm madly trying to finalise that. So, I don't plan to add any more new appliances (for v14.2). Although there is one other than has been hanging around for a little while (not as long as yours - but longer than I like), which I might squeeze in if I can. 

As soon as that's done, we'll be upgrading the base infrastructure to the new Debian release (Stretch). Once that's done, we'll then turn around and update all the current appliances to v15.0. I can then add any new v15.0 appliances you or others produce then.

If you want, you could make a start before then, but it's highly unlikely that we'll publish a v14.2 release. Having v14.2 based code would still give you a head start for v15.0, but will most likely need some tweaking.

So it's your call, but I think it would be best to wait.

**Comment by MPTMG** [comment]
I will wait. I will lay the groundwork for it now but will wait. I wasn't trying to slip in more work for you before 14.2, I just saw that you had integrated Mumble and will have ZoneMinder done shortly as well, so I figured I would tackle another appliance I find useful.

**Comment by JedMeister** [comment]
No worries. I don't want to put you off, I just wanted to be clear on where we're up to so you have a bit of context.

Mostly, I really just don't want to repeat the drawn out process that has happened with your 2 other submissions. It's a bit embarrassing really...! :smile: 

We love your contributions, despite the fact that it took us so long to release them. Actually part of the hold up has been that we've been refactoring things behind the scenes to try to do more rapid releases. I'm not yet convinced that it will make the releases tons quicker, but hopefully it should improve things...

**Comment by MPTMG** [comment]
Went ahead and finished the work on this and committed it to GitHub. Simple appliance to configure considering it has very few dependencies and inithooks are limited. FWIW here is the link to the repository https://github.com/MPTMG/Syncthing. I will do the artwork shortly. If anything else is needed let me know.

**Comment by JedMeister** [comment]
Awesome, thanks @MPTMG. We're still way behind on our next release, although the pace is starting to pick up. We still have a few major issues that need to be resolved before we can build a release candidate of Core (which will likely be the first public release of v15.0).

So unfortunately, I'm not sure when we'll get to release your new app, but thanks tons for your work.

I should post some instructions on how to start building for Stretch/v15.0 somewhere. It's actually fairly simple...

**Comment by JedMeister** [comment]
Great work @MPTMG 

I've just had a quick look at the code and here is some feedback/things we'll need to do prior to inclusion. I don't necessarily mean that you need to do them, but they will need to be done before we can add it and we're pretty snowed under ATM, so if you could have a look would be awesome!:

- [ ] [**`conf script`**](https://github.com/MPTMG/TKL-Syncthing/blob/master/conf.d/main):

  - [ ] good practice to include the `-y` switch when installing packages in conf scripts. Even if it doesn't need it ATM, future versions may require additional dependencies and without the `-y` switch the build will stall, waiting for confirmation. So it's a good habit to be in. :smile: 
  - [ ] package(s) installed from third party apt repo need to be pinned (for security). For example see [Mediaserver](https://github.com/turnkeylinux-apps/mediaserver) [`emby.pref`](https://github.com/turnkeylinux-apps/mediaserver/blob/master/overlay/etc/apt/preferences.d/emby.pref) - apt preferences overlay file for Emby

  - [ ] additionally, the apt repo should be secured via (new practice introduced for v15.0):

    - [ ] upstream apt gpg key should be saved in `overlay/usr/share/keyrings/APP_NAME.gpg`, i.e. `overlay/usr/share/keyrings/syncthing.gpg`
    - [ ] include `signed-by` line (pointing to gpg key) in the sources.list file. E.g. [`emby.list` overlay file](https://github.com/turnkeylinux-apps/mediaserver/blob/master/overlay/etc/apt/sources.list.d/emby.list)
    - rational and further info, see #1033 - strictly speaking these files don't necessarily need to be in overlay (could be in conf script), but most others are, so it's good for consistency.

- [ ] [**`changelog`**](https://github.com/MPTMG/TKL-Syncthing/blob/master/changelog) - remove and add single 15.0 entry (appliances should only have entries relating to their release - note that this is initial release).
- [ ] [**`readme`**](https://github.com/MPTMG/TKL-Syncthing/blob/master/README.rst) - adjust to bring formatting and layout into line with other appliance readmes (we semi-automatically process them to produce the appliance page text on the website)
- [ ] [**`appliance logo`**](https://github.com/MPTMG/TKL-Syncthing/blob/master/.art/logo.png) - it will need it's own logo/icon (currently still has Core logo/icon). The logo/icon template and some instructions can be found here: https://github.com/turnkeylinux/artwork
- [ ] **landing page** (i.e. tklweb-cp) **_or_ redirect** (via webserver, e.g. nginx) - as it's served on a non-standard port, IMO it would be of value to have either a landing page (can be handy to also provide doc links and third party app links) or redirect vanilla http/https (i.e. port 80/443) to the WebUI. TBH, my preference would be the former, but the latter is ok too.
- [ ] **screenshot(s)** - Ideally each appliance should include 3 screenshots. But if there aren't 3 screenshots of value, we can "recycle" common ones (such as Webmin) to make up the 3. I'd delete all 3 currently included as they'll need to be updated for v15.0 anyway. IMO at least one screenshot of the SyncThing webUI shoudl be included, perhaps 2 (or more) if they look interesting. If we include a landing page, then a screenshot of that would also be appropriate. Suggested workflow is noted in the [artwork repo readme](https://github.com/turnkeylinux/artwork/blob/master/README.rst) - more detailed notes/instructions are in the [docs](https://github.com/turnkeylinux/artwork/blob/master/docs/screenshots.rst).

**Comment by MPTMG** [comment]
I will complete the list. Only thing that seems remotely complex is pinning the third party apt repo, but I will  look at what was done for emby and try to reproduce for syncthing. Good work on the 15.0-rc1.

**Comment by JedMeister** [comment]
Thanks mate. Apologies that I didn't give this specific feedback sooner. Between support and trying to polish up v15.0, I've been snowed under...

Ideally, we probably should write up something like a "new appliance checklist" or similar. 

If you need any further assistance or guidance, get stuck, etc, please don't hesitate to ask.

Also, there's no super rush on this, just let me know when you're ready. I hope to start releasing v15.0 appliances within the next week, but same as v14.2, they'll be done in batches of ~20 so there'll be about 5 batches, a week or 2 apart. So there's plenty of time. And even if it isn't ready by the last batch, it's not that big a deal. Moving forward, I hope to do more regular builds of v15.x.

**Comment by MPTMG** [comment]
I took care of the conf script -y, added the gpg key via overlay, repo via overlay, and pinned the repo. 

If I blank out the change log it gives me a build error. What does it reference in the changelog for the build process? If I copy the default changelog from dev or core it builds fine. I have 100% confirmed it has something to do with the changelog. I was not under the impression the changelog could tamper with the build process.

I will update the readme and the art tomorrow. 

I do not see the value in including a web server with redirect or tklweb-cp (control panel) as I could easily just change the default hosting port on Syncthing to 80 or 443 and avoid the unnecessary packages. However, if someone is using Syncthing (bothered to read the docs) they will be expecting the webgui on 8384 so changing the default port for the Syncthing webgui would prove confusing for those users. Let me know if you would like the default port changed for the Syncthing webgui, but installing a web server nginx or lighttpd or apache along with php to essentially serve links via one php page to webmin/syncthing/terminal is not something I think has value to the end user. For the experienced user it would involve additional work to close the unnecessary open ports for the web server. You have full permission to give me hell if a ton of users are bothering you for support because they can't type :8384 for the syncthing webgui.

Will post code tomorrow after art and readme is done.

**Comment by JedMeister** [comment]
Hi @MPTMG - Great work.

Apologies if I wasn't clear. I didn't mean remove the changelog completely, I meant just have a single v15.0 entry.

The changelog needs to exist and it also needs to comply with Debian changelog policy. One way to ensure that it does is to copy/paste (or edit) an existing changelog entry and update it as appropriate.

Here's an example changelog for a new appliance called `new-appliance`:
```
turnkey-new-appliance-15.0 (1) turnkey; urgency=low

  * Initial appliance release based on LAMP appliance

  * Includes v1.0 of new-appliance, installed from upstream git repo

 -- Jeremy Davis <jeremy@turnkeylinux.org>  Mon, 02 Jul 2018 10:15:36 +1000
```

IMO a better way is to use the Debian developer `dch` tool to create a new entry (that's in the [`devscripts` package](https://packages.debian.org/stretch/devscripts)). I hope to one day make a smaller and easier to use tool for our purposes, but for now you can use it like this:

First set up your TKLDev with the appropriate env vars, here's mine (output of `cat /root/.bashrc.d/tkldev`):
```
export DEBFULLNAME="Jeremy Davis"
export DEBEMAIL="jeremy@turnkeylinux.org"
```
If you want the time/date to be in your local format (rather than UTC) set that up too (i.e. `dpkg-reconfigure tzdata`).

Then assuming that the changelog doesn't yet exist, within the appliance base directory (e.g. `/turnkey/fab/products/new-appliance`). Obviously adjust `APP_NAME` as appropriate:
```
APP_NAME=turnkey-new-appliance-15.0
dch --create --changelog changelog --package $APP_NAME --newversion 1 --force-distribution turnkey --urgency low
```

Which will produce this changelog:
```
turnkey-new-appliance-15.0 (1) UNRELEASED; urgency=low

  * turnkey

 -- Jeremy Davis <jeremy@turnkeylinux.org>  Mon, 02 Jul 2018 10:31:41 +1000
```

Then edit with your favourite editor and change `UNRELEASED` to `turnkey` and replace the `turnkey` entry with your desired entries.

FWIW, here's how to add a new entry to an existing changelog:
```
APP_NAME=turnkey-existing-appliance-15.0
dch --changelog changelog --release-heuristic log --package $APP_NAME --newversion 1 --force-distribution turnkey --urgency low
```
Apologies that I hadn't noted that somewhere previously. We should probably include that in the TKLDev docs...

**Comment by MPTMG** [comment]
Instructions were clear that the changelog had to be updated to reflect a new appliance, I just wasn't aware an improperly formatted changelog could cause problems to the build process and hence just blanked it out temporarily, while making other changes. I won't lie after making the minor changes for third party repo when it started throwing a build error I was completely baffled as to what the problem was considering I could not fathom it was the changelog.

Thank you for the detailed instructions on how to properly generate a changelog as it will be useful going forward.

**Comment by MPTMG** [comment]
This one should be ready. Take a look when the opportunity permits. Used core as template for readme adding just a few edits for Syncthing related details.

**Comment by JedMeister** [comment]
Looks pretty good from a glance. I've pulled it into the [library](https://github.com/turnkeylinux/syncthing).

I probably should have given you a little more guidance on the readme as it looks like you've put a ton of work in and while it looks fantastic, each appliance notes that it includes "all the standard features in TurnKey Core" so I probably could have saved you some effort there... So I'll probably pare it down a little to bring it inline with the others (as I noted, we process the text for the website).

The logo and screenshots look fantastic though and from a glance everything else looks fine. Legendary effort! :smile: 

I'll do a test build ASAP and we'll see how we go. I'll leave this issue open until I've actually got it built.

One other thing that I forgot to mention too, was we'll need a [TKLBAM profile](https://github.com/turnkeylinux/tklbam-profiles). If you look at the existing profiles, most of them are actually empty. That's because they all inherit from the [core profile](https://github.com/turnkeylinux/tklbam-profiles/blob/master/core).

If Syncthing doesn't need any extra data backed up beyond what is included by default then it can just have an empty profile too. But if it does need anything special added (or excluded), then we'll need to add that there.

From a glance, it'd be worth checking if `/var/lib/syncthing` exists and if it does, whats in there. FWIW, `/home/syncthing` won't need to be added though, as the core profile [includes `/home`](https://github.com/turnkeylinux/tklbam-profiles/blob/master/core#L7) by default.

FWIW here's a bit more info in case you need it:

Additional directories (and/or files) to include in the backup are simply noted, each one on a new line. Exclusions are added the same, except with a `-` in front. Please note that they are recursive. So to use the `mediaserver` appliance [profile](https://github.com/turnkeylinux/tklbam-profiles/blob/master/mediaserver) as an example:
```
/var/lib/emby-server
-/var/lib/emby-server/cache/temp
-/var/lib/emby-server/transcoding-temp
```
This will include all files and sub-directories (and files within them etc) in `/var/lib/emby-server`, except for `/var/lib/emby-server/cache/temp` (and it's contents - files and subdirs etc) and `/var/lib/emby-server/transcoding-temp`.

Any questions, please ask and I'll get back to you ASAP.

**Comment by MPTMG** [comment]
/var/lib/syncthing does not exist and syncthing runs on a user basis as a service. So everything related to an instance of syncthing will be in /home/user where in this case the user I chose by default was syncthing.

How do you want the TKLBAM profile handled? Do you want a fork with a pull request? Looks like every appliance I have contributed to will be covered under the core profile.

**Comment by JedMeister** [comment]
Ok cool. 

We need to explicitly add a profile file, even if it's empty. 

So long as I know that, it's probably easier for me to just add it. Which I've [just done](https://github.com/turnkeylinux/tklbam-profiles/commit/b7bb7772a6cda61b5ea7975478643e5e5dd4bf82) for syncthing. But I'll also need to do that for any other new apps too.

**Comment by JedMeister** [comment]
I just realised I'd forked this into the wrong repo (turnkeylinux instead of turnkeylinux-apps). I've just fixed that...

**Comment by JedMeister** [comment]
:point_up: @MPTMG - I've had a look at this one now and I'm pretty happy with it. I've opened a pull request (against the TurnKey fork; rather than pushing direct to the repo as i often do) so you can easily see what I've done and ask any questions you may have.

Firstly I'd like to say, nice work. Syncthing looks like an awesome piece of software and I think it'll make an awesome addition to the library! :smile: Now the (hopefully constructive) critique:

The naming convention for inithooks in `/usr/lib/inithooks/` is: `bin/APP_NAME.py` and `bin/40APP_NAME` and possibly `bin/20regen-APP_NAME-secrets` (where `APP_NAME` is the appliance name in lowercase). If additional hooks are required `APP_NAME` should be included in the name with the purpose as clear as possible (whilst trying to keep the name reasonably short).

As you can see I've refactored your inithooks and moved the initial config components into a (new) `firstboot.d/20regen-syncthing-secrets` script. That way the only processing the `bin/syncthing.py` script will need to do is adjust the password. It's not a big deal really, but it makes the inithooks a little easier to read (and maintain). It also means that if a user reruns  `bin/syncthing.py` it will only update the password (rather than regenerate everything) which is expected behaviour.

Ideally, it would probably be nice to complete the install during `conf.d/main` then reset the required "secrets" on firstboot. However, I can see why you didn't do that. As Syncthing auto-generates everything on first run, it makes sense to just let it take care of that itself.

The new `20regen-syncthing-secrets` should work whether it's already been run before or not. Its not perfect, but I think it should be pretty solid for the v15.0 release.

Please feel free comment or ask any questions you might have.

----

I'm not going to worry about it for now, but I have some ideas on how we could make it even better. 

I think that including a minimalist webserver and landing page (i.e. tklweb-cp) as we do with most other appliances (e.g. like we did with Zoneminder). Makes it easier for noob users to find the app without having to read the docs.

Alternatively we could leave Syncthing listening to localhost on plain http and hide the Web UI behind a reverse proxy (e.g. nginx)?!

That would also mean it would use the default system TLS/SSL certs with no adjustment.

Anther option would be to leave it self serving and allow it to leverage the default SSH keys that we already produce. 

I did have a quick look at that but by default only root has access to the default keys (in /etc/ssl/private), so we'd need to look at how we do that in a "best practice" way.

**Comment by MPTMG** [comment]
Thanks for the constructive criticism as the naming convention for hook will certainly make it easier to understand in the long run. I have not been using any convention and certainly additional appliances I have recently worked on will need this adjustment. I will endeavor to ensure the changes are made to reduce workload on you end in getting appliances integrated.

Good catch on enabling just password reset via the inithooks. I certainly did not pick that one up and was regenerating everything on turnkey-init.

I personally think it is cleaner to allow everything to be generated and started on first boot, so Syncthing can take care of the heavy lifting instead of trying to replace everything via scripts.

nginx reverse proxy is probably the cleanest and most secure of the application improvements that can be made. Uses the default TLS/SSL and provides the web interface on a default port. Is there any way to use iptables to forward requests from the standard syncthing port to 443? Such that if someone read the docs they will not be confused.

**Comment by JedMeister** [comment]
> Thanks for the constructive criticism as the naming convention for hook will certainly make it easier to understand in the long run. I have not been using any convention and certainly additional appliances I have recently worked on will need this adjustment. I will endeavor to ensure the changes are made to reduce workload on you end in getting appliances integrated.

Thanks man. That'd be awesome.

> I personally think it is cleaner to allow everything to be generated and started on first boot, so Syncthing can take care of the heavy lifting instead of trying to replace everything via scripts.

Yeah seeing as Syncthing does that itself so nicely, it seems silly to reinvent the wheel. The other upside is that if they change things, then we won't need to rewrite it! :smile: 

> nginx reverse proxy is probably the cleanest and most secure of the application improvements that can be made. Uses the default TLS/SSL and provides the web interface on a default port. Is there any way to use iptables to forward requests from the standard syncthing port to 443? Such that if someone read the docs they will not be confused.

Another thought I just had for a frontend, was we could use stunnel?! As that's already installed it wouldn't need anything additional. The downside with stunnel though is it only listens via TLS/SSL, so we couldn't redirect http to https (i.e. 80 to 443). I'm not sure, but perhaps stunnel could listen on both 443 and the default port though?

A brief google suggests that we probably could use iptables to redirect ports, but I'm not sure that's the best solution. Firstly, it's not enabled by default, so we'd need to change that. Also my brief reading suggests that we either need to know the IP address and/or the interface name for that to be reliable. We could work that stuff out with scripting, but it's likely to break in edge cases.

So thinking about it more, I reckon that Nginx would be the go. AFAIK, we should be able to have that also listen on the default Syncthing port and redirect to 443 too (as well as 80)!

**Comment by JedMeister** [comment]
I'm no expert when it comes to Nginx and I haven't tested, but I think this might work (or at least be a start):
```
# connect to Syncthing UI localhost on 8383 (non-standard Syncthing port)
upstream syncthing {
    server 127.0.0.1:8383;
}

# redirect http to https
server {
    listen 0.0.0.0:80;
    return 302 https://$host$request_uri;
}

# redirect 8384 (Syncthing UI default) to https
server {
    listen 0.0.0.0:8384;
    return 302 https://$host$request_uri;
}

server {

    listen 443 ssl;
    access_log /var/log/nginx/syncthing.log;
    error_log /var/log/nginx/syncthing_error.log;
    include /etc/nginx/include/ssl;

    location / {
        proxy_pass https://syncthing;
        proxy_redirect off;
    }
}
```
Note that we'd need to include the Nginx overlay (includes the SSL config) from common (i.e. in the Makefile, include `nginx` after `COMMON_OVERLAYS +=`. 

We'd also need to reconfigure Syncthing to use a different non-standard port (8383 in my above example).

**Comment by JedMeister** [comment]
FWIW, I merged my pull request for this one too. I think it's pretty much ready to go. I might have a real quick look at the Nginx stuff as posted above, but if I can't get it working fairly quickly, I might hold it off until v15.1.

**Comment by JedMeister** [comment]
Ok fixed a little bug and got the Nginx reverse proxy working. TBH, I spent way too much time on it, but got there in the end.

**Comment by MPTMG** [comment]
Awesome.

**Comment by MPTMG** [comment]
Didn't test yet but default port for sync thing is 8384 not 8383. I will test but after taking a quick look at code I do not see where you might have changed the default port to make it work with the current ngoni configure.

**Comment by MPTMG** [comment]
Nevermind I found the port change in regen-secrets hook. Excellent work by the way I am going to copy what you did for tvheadend web interface as it does not work via https by default but will now through the nginx proxy.

**Comment by MPTMG** [comment]
Might want to change the port to 8384 so it redirects via the proxy for users expecting web ui on 8384.

**Comment by JedMeister** [comment]
@MPTMG - glad you like it! :smile:

I think it's ready for a build and final test... :tada: 

I've just [tidied the Nginx config](https://github.com/turnkeylinux-apps/syncthing/commit/0c7f16e621988c8aa25ac5730d6f99a433c24748) up a little, removing some superfluous stuff. I also [adjusted the readme](https://github.com/turnkeylinux-apps/syncthing/commit/55c23f1a734c6a34a1c429d7324d95decd8ea5cf) so hopefully it's clear to users.

Re port 8383/8384, perhaps I'm missing something, but I'm pretty sure that it works as intended?!

Syncthing is [configured to listen on 8383](https://github.com/turnkeylinux-apps/syncthing/blob/f751624742da7bd438b74c7289f9ecc7902f33c6/overlay/usr/lib/inithooks/firstboot.d/20regen-syncthing-secrets#L45) (as you noted, a non standard port) but only to localhost (i.e. not available outside the server).

Nginx [listens on 8384](https://github.com/turnkeylinux-apps/syncthing/blob/f751624742da7bd438b74c7289f9ecc7902f33c6/overlay/etc/nginx/sites-available/syncthing#L9) (standard Syncthing port) and [redirects that to HTTPS](https://github.com/turnkeylinux-apps/syncthing/blob/f751624742da7bd438b74c7289f9ecc7902f33c6/overlay/etc/nginx/sites-available/syncthing#L10) (i.e. port 443). Nginx is also [listening on port 443](https://github.com/turnkeylinux-apps/syncthing/blob/f751624742da7bd438b74c7289f9ecc7902f33c6/overlay/etc/nginx/sites-available/syncthing#L15) (i.e. HTTPS) which is a [reverse proxy for localhost:8383](https://github.com/turnkeylinux-apps/syncthing/blob/f751624742da7bd438b74c7289f9ecc7902f33c6/overlay/etc/nginx/sites-available/syncthing#L26) (i.e. Syncthing).

The only other thing that might have been nice is if we could listen to both http and https traffic on port 8384 and redirect both back to 443. Unfortunately though Nginx isn't capable of that. It can only listen for either plain http OR https (encrypted http) traffic. To accept both on a single port, we'd need some sort of network multiplexer listening on 8284, which then hands over to Nginx. Whilst it'd be nice, IMO it would be significant overkill for a simple convenience like that.

**Comment by JedMeister** [comment]
Actually, I stand corrected on the whole "Nginx isn't capable of [listening for both http and https traffic on a single port]" thing! See [this](https://serverfault.com/questions/47876/handling-http-and-https-requests-using-a-single-port-with-nginx/664936#664936) SeverFault answer.

I've just tested [it](https://github.com/turnkeylinux-apps/syncthing/blob/4d527959a083c570932aba56e8ea7820421af5d9/overlay/etc/nginx/sites-available/syncthing#L9-L11) and it works! :tada: :smile:

**Comment by MPTMG** [comment]
That is a clever solution. Listen SSL and redirect error.

**Comment by JedMeister** [comment]
Yeah that's what I thought... :smile:

**Comment by JedMeister** [comment]
Turns out that the Syncthing peeps don't recommend using the `proxy_set_header Host localhost;` within Nginx config (with Syncthing bound to localhost). See the discussion over on https://github.com/syncthing/docs/issues/401

Instead it should be setting `proxy_set_header Host $host;` (so Syncthing is aware of where the connection is really coming from). However, when I tried that, with Syncthing bound to localhost, I got a `host check error` (because the connection isn't coming from localhost). The workaround for that is to set `<insecureSkipHostcheck>true</insecureSkipHostcheck>` (within the `<gui .. > </gui>` section of the Syncthing `config.xml`). So perhaps that's what we should do instead?

Alternatively, we could allow Syncthing to listen globally (i.e. `0.0.0.0/0`), but IMO that makes the whole idea of a reverse proxy redundant!

So I'm not really sure of the best approach...

If we ditch the idea of a reverse proxy, we could just set up Nginx with a redirect from ports 80/443 to 8384? Or a landing page (i.e. tklweb-cp)...

Unfortunately though, that gets us back to where we were with the lack of shared whole system SSL/TLS certs. We could allow access to the default certs (in `/etc/ssl/private` - or copy them across etc). But as we currently lock them down to be only readable by root, we'd need to relax that restriction (Syncthing runs as a non-root user). I'm not totally convinced that's a good idea...

So in summary, I feel a little stuck. None of the options seem ideal.

**Comment by MPTMG** [comment]
I would just let syncthing generate it's own cert. Redirect 80 and 443 to 8384 via nginx. 

My justification is that no matter how certs are handled be it syncthing or global certs the syncthing user still needs access to the certs to operate. Nginx proxy to idiot proof the connection to syncthing.

In the end this is the most secure and simple. We do not break best practice for system wide certs and syncthing is operating as it would in a default install.

Hence we have not modified the security of either the appliance or inherent security in syncthing. All of the other options involve a trade off where we decrease system security to increase syncthing security or increase system security while decreasing syncthing security. Best option to me is leave it alone and handle them separately.

**Comment by JedMeister** [comment]
FWIW I'm currently leaning towards setting `<insecureSkipHostcheck>true</insecureSkipHostcheck>` for the GUI.

**Comment by MPTMG** [comment]
I would not change anything. The guys at syncthing I think misunderstood your question. You are only proxying the webgui, which by default only listens on localhost. You are not interfering with syncthing in the least as far as how it functions. You are simply allowing remote access to the webgui via a nginx proxy. Sure allowing remote access has it's security implications, but no more than having syncthing listen on 0.0.0.0:8384. Which via proxy and it listening on 0.0.0.0:8384 are the only two ways to access it remotely.

So in essence operating without remote access syncthing only ever sees access from localhost. I do not see how this could have additional security implications, but I have been wrong before.

**Comment by JedMeister** [comment]
> Sure allowing remote access has it's security implications, but no more than having syncthing listen on 0.0.0.0:8384. Which via proxy and it listening on 0.0.0.0:8384 are the only two ways to access it remotely.

I'm inclined to agree, although they're probably right that it's better to explicitly/trasnparently forward the traffic (via Nginx) and disable Syncthing's Hostcheck (i.e. `<insecureSkipHostcheck>true</insecureSkipHostcheck>`).

But just in case we're missing something, I've asked them for clarity on that point. I've also given more details on the issue that I'm trying to solve (i.e. global TLS certs).

**Comment by JedMeister** [comment]
This one is now done too :tada: :clinking_glasses: 

https://www.turnkeylinux.org/syncthing

**Comment by JedMeister** [comment]
PS thanks again for your work on this @MPTMG - we've got a bit of positive feedback on this one already! :smile:
